### ISSUE TITLE
Lazy loading

### ISSUE THREAD

**Issue by aidevnn** [SciSharp/NumSharp#326]
Hi
Your lib NumSharp was inspired me and i tried to write a other approach of NumPy more easy to use.
The big idea is the Lazy Loading for the front end and this is my repo.
https://github.com/aidevnn/DesertLand

My lib isnt optimized for speed at this time, but i can improve it with ease by rewritting NDarray backend class and methods WITHOUT changing nothing from the NDview struct frontend which use the lazy loading.

Actually your Backend is very very powerfull and maybe the frontend can be also improved with Lazy Loading.

Best regards.

**Comment by Oceania2018** [comment]
@aidevnn Join us, build Deep Learning library based on current work.

**Comment by aidevnn** [comment]
I started to fork your repo, it takes me some month to understand in deep your great work. 
Best regards.

**Comment by Nucs** [comment]
Hey @aidevnn, Thank you.
A. Are you aware we are rewriting our backend in a [separate branch](https://github.com/SciSharp/NumSharp/tree/unmanaged-bytes-storage)
B. Can you elaborate more about what do you mean by a front-end lazy loading?
If I understand correctly, lazy-loading is something that numpy does not implement therefore if you wish to contribute a lazy-loading front-end when NumSharp is the backend - it'll have to be in a seperate project.

**Comment by aidevnn** [comment]
Hello @Nucs 

A. Ok a devel branch is more collaborative than a fork

B. Lazy Loading can separate the SharpNum library onto frontend with simple to write specification which will be used by the rest of SciSharp ML/AI framework and the pure SharpNum backend for fast and speed computation. The SciSharp ML/AI team can write their own unit test to integrate SharpNum progression.
Lazy Loading is a low level design pattern wich allow to build complex expressions and he acts sometime like <<proxy>>. It is also a possible abstraction for integrating multiple different backend but it isnt a good idea at this moment for the framework SciSharp.

For example, the expression 
NDproxy d = a+2*b+np.log(c);
NDarray e = d;
will only be evaluated during the implicit type conversion.
I am a self taugh and i discovered the lazy loading just now to reduce computation. I have some intuition about some other usage of the Lazy Loading, but i cannot explain it.

**Comment by Oceania2018** [comment]
@aidevnn Is `NDProxy` like a kind of Expression/ Func/ Delegate ? It can be executed after all operations have been decided ?

**Comment by aidevnn** [comment]
Yes it is a struct that can be defined like this
NDproxy a = new NDproxy(()=>ndArray0)
Its contains only on property : Func<NDarray> fnc, which can be called explicitly 
NDarray b = a.fnc() during internal computation of the backend 
or implicitly during type conversion by the frontends users.

**Comment by Nucs** [comment]
@aidevnn 
A. You can still use that branch from a fork
B. Thanks for laying it out for me, I am too a self taught.
Honestly - there is no fast way to stack lazy loading ops. Any basic implementation in C++ will surpass C#'s performance by far.
Your best luck to put effort is [Tensorflow.NET](https://github.com/SciSharp/TensorFlow.NET) since tensorflow holds a somewhat similar API to numpy and is capable of computation on a GPU and it provides a pretty impressive performance.

if you do insist on writing it on your own then you better get started with IL generation that will link the math ops together.
Mainly because every math-op method you delegate to the next math-op method is adding more overhead than you would think.

**Comment by aidevnn** [comment]
@Nucs  you are right! Lazy Loading and some stacks operations because its a kind of managing expression pattern.
Managing code have advantage for simplifying coding and maintenance, but with loss of speed and performance. Its a tradeoff.

**Comment by aidevnn** [comment]
@Nucs The idea behind the lazy loading is to introduce step by step some symbolic neural network algorithms with the numerical connexionnist algorithms.
For beginning, a precompilation of all requirement like shape / strides / slices can be done without recreating them each time and its my first objectif. I will let know my progress

**Comment by aidevnn** [comment]
In the expression
```
NDview<double> b = 5*np.log(a+1)+3
```
In the above expression, my actual code and symbolic approch reduce rawdata access only to one time. Also in a more complexe expression, the lazy loading reduce the data access to the minimum effort.

```
NDview<double> b = 5*np.log(a+1)+ np.log(a+2)
```
A NDarray appears two time, and there will be only two time rawdata access at all. But i am trying to solve some performances problems.

We can say that a lazy loading pattern can improve the performance in theory but it introduces some other stack call problems. I will continue to search how to take advantage of it.

**Comment by aidevnn** [comment]
I discovered in this article 
https://techdecoded.intel.io/resources/parallelism-in-python-directing-vectorization-with-numexpr/#gs.63gvwl

Numexpr : it is very usefull and it already use lazy-loading and a lot of symbolics optimisations to enhance performance.

**Comment by Nucs** [comment]
@aidevnn, partial lazy loading shipped in **0.70.0** (#628): `np.load(mmap_mode:)` returns a zero-copy memory-mapped array, and `NpzFile` is lazy + cached:

```csharp
// read-only mmap - the file is mapped, not read into memory up front
var a = np.load("big.npy", mmap_mode: "r");
Console.WriteLine(a.flags.writeable);   // False

// "r+" maps read-write (writes flush to disk); "c" is copy-on-write
```

Closing (a full lazy-evaluation graph is a separate, larger effort).
